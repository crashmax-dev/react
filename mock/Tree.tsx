import React from 'react'

export function App(props: React.PropsWithChildren) {
  return <div aria-label="app">{props.children}</div>
}

export function Layout(props: React.PropsWithChildren) {
  return <div aria-label="layout">{props.children}</div>
}

type StateProps = React.PropsWithChildren<{
  initialValue: string
}>

export function State(props: StateProps) {
  return (
    <div aria-label="state">
      <h1 aria-label="state-initial-value">{props.initialValue}</h1>
      {props.children}
    </div>
  )
}
