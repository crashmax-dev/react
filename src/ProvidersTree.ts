import { createElement } from 'react'

type Component = React.FunctionComponent<any>
type Props<T extends Component> = Omit<Parameters<T>[0], 'children'>
type Provider<T extends Component> = [Component, Props<T>?]

export function ProvidersTree(props: {
  children?: JSX.Element
  providers: (
    wrapper: <T extends Component>(
      component: T,
      props?: Props<T>
    ) => Provider<T>
  ) => Provider<Component>[]
}) {
  const providers = props.providers(componentWrapper)
  let currentChildren = props.children!

  for (let i = providers.length - 1; i >= 0; i--) {
    const [component, props = {}] = providers[i]!
    currentChildren = createElement(component, props, currentChildren)
  }

  return currentChildren
}

function componentWrapper<T extends Component>(
  component: T,
  props?: Parameters<T>[0]
): Provider<T> {
  return [component, props]
}
