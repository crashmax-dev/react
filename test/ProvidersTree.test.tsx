import { App, Layout, State } from '../mock/Tree.js'
import { ProvidersTree } from '../src/ProvidersTree.js'
import { render, screen } from './utils.js'

describe('ProvidersTree', () => {
  it('should be defined', () => {
    expect(ProvidersTree).toBeDefined()
  })

  it('should be render component tree with children', () => {
    render(
      <ProvidersTree
        providers={(wrapper) => [
          wrapper(State, { initialValue: 'hello' }),
          wrapper(Layout)
        ]}
      >
        <App />
      </ProvidersTree>
    )
    expect(screen.getByLabelText('state')).toMatchSnapshot()
  })

  it('should be render component tree without children', () => {
    render(
      ProvidersTree({
        providers(wrapper) {
          return [
            wrapper(State, { initialValue: 'hello' }),
            wrapper(Layout),
            wrapper(App)]
        }
      })
    )

    expect(screen.getByLabelText('state')).toMatchSnapshot()
  })
})
