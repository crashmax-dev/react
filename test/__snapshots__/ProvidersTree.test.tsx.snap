// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProvidersTree > should be render component tree with children 1`] = `
<div
  aria-label="state"
>
  <h1
    aria-label="state-initial-value"
  >
    hello
  </h1>
  <div
    aria-label="layout"
  >
    <div
      aria-label="app"
    />
  </div>
</div>
`;

exports[`ProvidersTree > should be render component tree without children 1`] = `
<div
  aria-label="state"
>
  <h1
    aria-label="state-initial-value"
  >
    hello
  </h1>
  <div
    aria-label="layout"
  >
    <div
      aria-label="app"
    />
  </div>
</div>
`;
